### PR TITLE
Add key to BottomNavigationBarItem

### PR DESCRIPTION
**Pull request by Gibbo97**
Pass key into _BottomNavigationTile

Adding a optional key parameter to BottomNavigationBarItem to be passed through to _BottomNavigationTile.
This will allow easier testing in some scenarios, and fix the splash appearing on the wrong tile.

https://github.com/flutter/flutter/issues/139615
https://github.com/flutter/flutter/issues/34833

### CODE REVIEW

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart:34]: Instead of referencing a private API here, can you explain what the key is used for, and an example use case (doesn't have to be a code example, just a use case, although I wouldn't turn a code example down. :-) ).

Comment (Gibbo97) [reply packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart]: thanks for the review, have updated. Hope this is what you were after :)

**Comment by bleroux** [comment]
About the 'Linux analyze' step failing, the related error message that will help you getting this step green is :

```
Analyzing flutter...                                            

   info • Local variables should be final • packages/flutter/test/material/bottom_navigation_bar_test.dart:3129:5 • prefer_final_locals
```

**Comment by FDuhen** [comment]
Thank you for your work, waiting with impatience for this PR to merged

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart:37]:
```suggestion
  /// This allows the identification of different [BottomNavigationBarItem]s through their keys.
```

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart:40]:
```suggestion
  /// each item a key will allow the inkwell / splash animation to be correctly positioned.
```

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

(approved, modulo the comments below)

**Comment by gspencergoog** [comment]
@Gibbo97 are you able to resolve the conflicts here and bring the PR up to date?

**Comment by Gibbo97** [comment]
@gspencergoog ready to be merged

**Comment by gspencergoog** [comment]
Okay, great, we'll need a second reviewer (I'll find someone) and then we can merge.

File: packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart
Comment (QuncCccccc) [inline packages/flutter/lib/src/widgets/bottom_navigation_bar_item.dart:34]: nit:
```suggestion
  /// A key to be passed through to the resultant widget.
```

**Review by QuncCccccc — APPROVED** [review]
LGTM:) Thanks for the contribution!

**Pull request merged**
